jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenShortestPathIterator.java

```diff
@@ -152,18 +163,77 @@ public YenShortestPathIterator(
             throw new IllegalArgumentException("Graph should contain sink vertex!");
         }
         this.sink = sink;
+        this.pathValidator = pathValidator;
         Objects.requireNonNull(heapSupplier, "Heap supplier cannot be null");
         this.resultList = new ArrayList<>();
         this.candidatePaths = heapSupplier.get();
-        this.deviations = new HashMap<>();
-        this.weightsFrequencies = new HashMap<>();
+        this.firstDeviations = new HashMap<>();
+        this.lastDeviations = new HashMap<>();
 
         GraphPath<V, E> shortestPath = DijkstraShortestPath.findPathBetween(graph, source, sink);
+
         if (shortestPath != null) {
-            candidatePaths.insert(shortestPath.getWeight(), shortestPath);
-            deviations.put(shortestPath, source);
-            weightsFrequencies.put(shortestPath.getWeight(), 1);
+            V lastValidDeviation = getLastValidDeviation(shortestPath, source);
```

Comment:
This `if` statement is a heavy computation and should be moved to the `hasNext` or `next` methods. The reason is that the user doesn't expect the constructor to do potentially all the computation.